### PR TITLE
Avoid rounding issues in MLOG instruction

### LINKED ISSUE #150: Logarithm instruction `mlog` needs discussion
I would expect that `assert(1000.logarithm(10) == 3);` would pass, but it does not.
Instead, logging `1000.logarithm(10)` returns `2`,

The testing I've done uses this implementation in sway:
```
impl Logarithm for u64 {
    fn logarithm(self, base: Self) -> Self {
        asm(r1: self, r2: base, r3) {
            mlog r3 r1 r2;
            r3: Self
        }
    }
}
```
Note that other cases are working as expected, i.e:
```
assert(4.logarithm(2) == 2);
assert(16.logarithm(2) == 4);
assert(9.logarithm(3) == 2);
assert(100.logarithm(10) == 2);
```

### PR DESCRIPTION
**Pull request by Dentosal**
Closes #150.

See https://github.com/FuelLabs/fuel-vm/issues/150#issuecomment-1288797787 for detailed explanation and benchmark.

This is a minor breaking change: it fixes a bug, but it's unlikely that anything is depending on the broken behavior.

This contains a performance regression, but correctness is more important here. ~~In any case, this has to be benchmarked and gas cost adjusted.~~ Initial benchmarking shows no significant impact. Either the benchmark needs to be changed to add the edge cases, or the pricing is good enough already.

### CODE REVIEW

File: src/interpreter/executors/instruction.rs [hunk 1]
```diff
@@ -496,5 +490,39 @@ where
     }
 }
 
+/// Computes logarithm for given exponent and base.
+/// Diverges when exp == 0 or base <= 1.
+#[inline(always)] // Force copy of each invocation for optimization, see comments below
+const fn _unchecked_ilog_inner(exp: Word, base: Word) -> u32 {
```
Comment (Voxelot) [inline src/interpreter/executors/instruction.rs:496]: Maybe include a reference to where this code was sourced from?

File: src/interpreter/executors/instruction.rs [hunk 2]
```diff
@@ -496,5 +490,39 @@ where
...
+
+/// Logarithm for given exponent and an arbitrary base, rounded
+/// rounded down to nearest integer value.
+///
+/// Must not be called if the exponent == 0, or if the base <= 1,
+/// as that causes either division by zero, or an infinite loop.
+///
+/// TODO: when https://github.com/rust-lang/rust/issues/70887 is stabilized,
+/// consider using that instead.
+const fn unchecked_ilog(exp: Word, base: Word) -> u32 {
```
Comment (Voxelot) [inline src/interpreter/executors/instruction.rs:514]: Why do we want this to be unchecked? Is there an efficient way to make this checked?

Comment (Voxelot) [reply src/interpreter/executors/instruction.rs:514]: see: https://github.com/FuelLabs/fuel-vm/issues/170

Comment (Dentosal) [reply src/interpreter/executors/instruction.rs:514]: This is unchecked due to the check already being done on callsite, where we use `alu_error` to check for the invariants. However, the compiler should be able to eliminate the `.expect(...)`, so I've changed this to `checked_` instead.

File: src/interpreter/executors/instruction.rs (hunk 1 above)
Comment (Dentosal) [reply src/interpreter/executors/instruction.rs:496]: Done

Comment (Voxelot) [reply src/interpreter/executors/instruction.rs:496]: Not seeing any changes, was a push missed somewhere?

Comment (Dentosal) [reply src/interpreter/executors/instruction.rs:507]: Yep, fixed now

**Review by Voxelot — APPROVED** [review]

**Pull request merged**